docs/reference/advanced_configuration.rst
Advanced Configuration
======================

Full configuration options:

.. code-block:: yaml

    sonata_media:
        db_driver: doctrine_orm
        class:
            media: App\Entity\SonataMediaMedia
            gallery: App\Entity\SonataMediaGallery
            gallery_item: App\Entity\SonataMediaGalleryItem
            category: null # App\Entity\SonataClassificationCategory if exists

        force_disable_category: false # true, if you really want to disable the relation with category

        default_context: default
        admin_format: { width: 200, quality: 90, format: 'jpg' }
        contexts:
            default:  # the default context is mandatory
                download:
                    strategy: sonata.media.security.superadmin_strategy
                providers:
                    - sonata.media.provider.dailymotion
                    - sonata.media.provider.youtube
                    - sonata.media.provider.image
                    - sonata.media.provider.file

                formats:
                    small: { width: 100, quality: 70 }
                    big: { width: 500, quality: 70, resizer: sonata.media.resizer.square }
                    # You can pass through any custom option to resizer by using the resizer_options key
                    icon: { width: 32, quality: 70, resizer: your.custom.resizer, resizer_options: { custom_crop: true } }

            tv:
                download:
                    strategy: sonata.media.security.superadmin_strategy
                providers:
                    - sonata.media.provider.dailymotion
                    - sonata.media.provider.youtube
                    - sonata.media.provider.video

                formats:
                    cinema: { width: 1850, height: 768 }
                    grandmatv: { width: 640, height: 480 }

            news:
                download:
                    strategy: sonata.media.security.superadmin_strategy
                providers:
                    - sonata.media.provider.dailymotion
                    - sonata.media.provider.youtube
                    - sonata.media.provider.image
                    - sonata.media.provider.file

                formats:
                    small: { width: 150, quality: 95 }
                    big: { width: 500, quality: 90 }

        cdn:
            server:
                path: /uploads/media

            panther:
                path: http://domain.pantherportal.com/uploads/media
                site_id:
                password:
                username:

            cloudfront:
                path: https://xxxxxxxxxxxxxx.cloudfront.net/uploads/media
                distribution_id:
                key:
                secret:
                region:
                version:

            fallback:
                primary: sonata.media.cdn.panther
                fallback: sonata.media.cdn.server

        providers:
            file:
                service: sonata.media.provider.file
                resizer: null
                filesystem: default.storage # from your Flysystem configuration
                cdn: sonata.media.cdn.server
                generator: sonata.media.generator.default
                thumbnail: sonata.media.thumbnail.format
                allowed_extensions: ['pdf', 'txt', 'rtf', 'doc', 'docx', 'xls', 'xlsx', 'ppt', 'pttx', 'odt', 'odg', 'odp', 'ods', 'odc', 'odf', 'odb', 'csv', 'xml']
                allowed_mime_types: ['application/pdf', 'application/x-pdf', 'application/rtf', 'text/html', 'text/rtf', 'text/plain']

            image:
                service: sonata.media.provider.image
                resizer: sonata.media.resizer.simple # sonata.media.resizer.square, sonata.media.resizer.crop
                filesystem: default.storage # from your Flysystem configuration
                cdn: sonata.media.cdn.server
                generator: sonata.media.generator.default
                thumbnail: sonata.media.thumbnail.format
                allowed_extensions: ['jpg', 'png', 'jpeg']
                allowed_mime_types: ['image/pjpeg', 'image/jpeg', 'image/png', 'image/x-png']

            youtube:
                service: sonata.media.provider.youtube
                resizer: sonata.media.resizer.simple
                filesystem: default.storage # from your Flysystem configuration
                cdn: sonata.media.cdn.server
                generator: sonata.media.generator.default
                thumbnail: sonata.media.thumbnail.format
                html5: false

            dailymotion:
                service: sonata.media.provider.dailymotion
                resizer: sonata.media.resizer.simple
                filesystem: default.storage # from your Flysystem configuration
                cdn: sonata.media.cdn.server
                generator: sonata.media.generator.default
                thumbnail: sonata.media.thumbnail.format

        http:
            client: sonata.media.http.base_client # You need symfony/http-client for this
            message_factory: sonata.media.http.base_message_factory # You need nyholm/psr7 for this

        messenger:
            enabled: false
            generate_thumbnails_bus: messenger.default_bus


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [4.20.0](https://github.com/sonata-project/SonataMediaBundle/compare/4.19.0...4.20.0) - 2025-12-19
### Added
- [[#2507](https://github.com/sonata-project/SonataMediaBundle/pull/2507)] Symfony 8 support ([@VincentLanglet](https://github.com/VincentLanglet))

### Removed
- [[#2505](https://github.com/sonata-project/SonataMediaBundle/pull/2505)] Support for PHP 8.1 ([@dmaicher](https://github.com/dmaicher))
- [[#2505](https://github.com/sonata-project/SonataMediaBundle/pull/2505)] Support for Symfony 7.1 and 7.2 ([@dmaicher](https://github.com/dmaicher))

## [4.19.0](https://github.com/sonata-project/SonataMediaBundle/compare/4.18.0...4.19.0) - 2025-10-18
### Added
- [[#2500](https://github.com/sonata-project/SonataMediaBundle/pull/2500)] Added `media.php` routing config ([@dmaicher](https://github.com/dmaicher))

### Deprecated
- [[#2500](https://github.com/sonata-project/SonataMediaBundle/pull/2500)] Deprecated `media.xml` routing config ([@dmaicher](https://github.com/dmaicher))

## [4.18.0](https://github.com/sonata-project/SonataMediaBundle/compare/4.17.0...4.18.0) - 2025-09-23
### Added
- [[#2497](https://github.com/sonata-project/SonataMediaBundle/pull/2497)] Support for `doctrine/persistence` 4.0. ([@jordisala1991](https://github.com/jordisala1991))

## [4.17.0](https://github.com/sonata-project/SonataMediaBundle/compare/4.16.0...4.17.0) - 2025-09-15
### Added
- [[#2492](https://github.com/sonata-project/SonataMediaBundle/pull/2492)] Basic support for using Azure Blob Storage as filesystem ([@jorrit](https://github.com/jorrit))

## [4.16.0](https://github.com/sonata-project/SonataMediaBundle/compare/4.15.0...4.16.0) - 2025-02-24
### Changed
- [[#2476](https://github.com/sonata-project/SonataMediaBundle/pull/2476)] Use BIGINT for Media::$contentSize ([@OskarStark](https://github.com/OskarStark))

## [4.15.0](https://github.com/sonata-project/SonataMediaBundle/compare/4.14.1...4.15.0) - 2024-11-14
### Added
- [[#2466](https://github.com/sonata-project/SonataMediaBundle/pull/2466)] Slovak translations ([@fbuchlak](https://github.com/fbuchlak))

### Changed
- [[#2465](https://github.com/sonata-project/SonataMediaBundle/pull/2465)] Czech translations updated (key order, incomplete values, missing keys) ([@fbuchlak](https://github.com/fbuchlak))

## [4.14.1](https://github.com/sonata-project/SonataMediaBundle/compare/4.14.0...4.14.1) - 2024-09-12
### Fixed
- [[#2460](https://github.com/sonata-project/SonataMediaBundle/pull/2460)] Fixed the callback transformers of media/gallery block services. ([@lemorragia](https://github.com/lemorragia))

## [4.14.0](https://github.com/sonata-project/SonataMediaBundle/compare/4.13.0...4.14.0) - 2024-08-27
### Added
- [[#2461](https://github.com/sonata-project/SonataMediaBundle/pull/2461)] Support for psr/http-message 2 ([@VincentLanglet](https://github.com/VincentLanglet))

### Fixed
- [[#2455](https://github.com/sonata-project/SonataMediaBundle/pull/2455)] Symfony 7.1 deprecation about `Symfony\Component\HttpKernel\DependencyInjection\Extension` usage ([@VincentLanglet](https://github.com/VincentLanglet))
- [[#2452](https://github.com/sonata-project/SonataMediaBundle/pull/2452)] Remove format restriction on `FileThumbnail` service ([@VincentLanglet](https://github.com/VincentLanglet))

## [4.13.0](https://github.com/sonata-project/SonataMediaBundle/compare/4.12.0...4.13.0) - 2024-03-17
### Added
- [[#2448](https://github.com/sonata-project/SonataMediaBundle/pull/2448)] Compatibility with ORM 3 ([@dmaicher](https://github.com/dmaicher))

## [4.12.0](https://github.com/sonata-project/SonataMediaBundle/compare/4.11.1...4.12.0) - 2024-03-08
### Changed
- [[#2444](https://github.com/sonata-project/SonataMediaBundle/pull/2444)] Throw TransformationFailedException when reverse transformation fails in ProviderDataTransformer. ([@jorrit](https://github.com/jorrit))

## [4.11.1](https://github.com/sonata-project/SonataMediaBundle/compare/4.11.0...4.11.1) - 2024-02-18
### Fixed
- [[#2442](https://github.com/sonata-project/SonataMediaBundle/pull/2442)] Resolve "Admin has no subject" error from the getMediaBuilder ([@VincentLanglet](https://github.com/VincentLanglet))
- [[#2442](https://github.com/sonata-project/SonataMediaBundle/pull/2442)] Resolve "Admin has no subject" error from the getGalleryBuilder ([@VincentLanglet](https://github.com/VincentLanglet))

## [4.11.0](https://github.com/sonata-project/SonataMediaBundle/compare/4.10.5...4.11.0) - 2023-12-15
### Added
- [[#2430](https://github.com/sonata-project/SonataMediaBundle/pull/2430)] Support for packages from `symfony/*` 7.x ([@phansys](https://github.com/phansys))

## [4.10.5](https://github.com/sonata-project/SonataMediaBundle/compare/4.10.4...4.10.5) - 2023-11-21
### Changed
- [[#2351](https://github.com/sonata-project/SonataMediaBundle/pull/2351)] How `picture` element twig options are merged, making them work same way as `srcset` ([@haivala](https://github.com/haivala))

### Fixed

src/DependencyInjection/Configuration.php
<?php

declare(strict_types=1);

/*
 * This file is part of the Sonata Project package.
 *
 * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Sonata\MediaBundle\DependencyInjection;

use Imagine\Image\ManipulatorInterface;
use Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition;
use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;

/**
 * This is the class that validates and merges configuration from your app/config files.
 *
 * To learn more see {@link http://symfony.com/doc/current/cookbook/bundles/extension.html#cookbook-bundles-extension-config-class}
 */
final class Configuration implements ConfigurationInterface
{
    private const DB_DRIVERS = ['doctrine_orm', 'doctrine_mongodb', 'no_driver'];

    /**
     * @psalm-suppress UndefinedInterfaceMethod
     */
    public function getConfigTreeBuilder(): TreeBuilder
    {
        $treeBuilder = new TreeBuilder('sonata_media');
        $rootNode = $treeBuilder->getRootNode();

        $rootNode
            ->children()
                ->scalarNode('db_driver')
                    ->defaultValue('no_driver')
                    ->info('Choose persistence mechanism driver from the following list: "doctrine_orm", "doctrine_mongodb"')
                    ->validate()
                        ->ifNotInArray(self::DB_DRIVERS)
                        ->thenInvalid('SonataMediaBundle - Invalid db driver %s.')
                    ->end()
                ->end()
                ->scalarNode('default_context')->isRequired()->end()
                ->booleanNode('force_disable_category')
                    ->info('true IF you really want to disable the relation with category')
                    ->defaultFalse()
                ->end()
                ->arrayNode('admin_format')
                    ->info('Configures the thumbnail preview for the admin')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->scalarNode('width')->defaultValue(200)->end()
                        ->scalarNode('height')->defaultValue(null)->end()
                        ->scalarNode('quality')->defaultValue(90)->end()
                        ->scalarNode('format')->defaultValue('jpg')->end()
                        ->scalarNode('constraint')->defaultValue(true)->end()
                        ->scalarNode('resizer')->defaultNull()->end()
                        ->arrayNode('resizer_options')
                            ->info('options directly passed to selected resizer. e.g. {use_crop: true, crop_gravity: center}')
                            ->defaultValue([])
                            ->useAttributeAsKey('name')
                            ->prototype('scalar')
                        ->end()
                    ->end()
                ->end()
            ->end();

        $this->addContextsSection($rootNode);
        $this->addCdnSection($rootNode);
        $this->addFilesystemSection($rootNode);
        $this->addProvidersSection($rootNode);
        $this->addModelSection($rootNode);
        $this->addHttpClientSection($rootNode);
        $this->addResizerSection($rootNode);
        $this->addAdapterSection($rootNode);
        $this->addMessengerSection($rootNode);

        return $treeBuilder;
    }

    /**
     * @psalm-suppress UndefinedInterfaceMethod
     */
    private function addContextsSection(ArrayNodeDefinition $node): void
    {
        $node
            ->children()
                ->arrayNode('contexts')
                    ->useAttributeAsKey('id')
                    ->prototype('array')
                        ->children()
                            ->arrayNode('download')
                                ->addDefaultsIfNotSet()
                                ->children()
                                    ->scalarNode('strategy')->defaultValue('sonata.media.security.superadmin_strategy')->end()
                                ->end()
                            ->end()
                            ->arrayNode('providers')
                                ->prototype('scalar')
                                    ->defaultValue([])
                                ->end()
                            ->end()
                            ->arrayNode('formats')
                                ->useAttributeAsKey('id')
                                ->prototype('array')
                                    ->children()
                                        ->integerNode('width')->defaultValue(null)->end()
                                        ->integerNode('height')->defaultValue(null)->end()
                                        ->integerNode('quality')->defaultValue(80)->end()
                                        ->scalarNode('format')->defaultValue('jpg')->end()
                                        ->booleanNode('constraint')->defaultValue(true)->end()
                                        ->scalarNode('resizer')->defaultNull()->end()
                                        ->arrayNode('resizer_options')
                                            ->info('options directly passed to selected resizer. e.g. {use_crop: true, crop_gravity: center}')
                                            ->defaultValue([])
                                            ->useAttributeAsKey('name')
                                            ->prototype('scalar')
                                        ->end()
                                    ->end()
                                ->end()
                            ->end()
                        ->end()
                    ->end()
                ->end()
            ->end();
    }

    /**
     * @psalm-suppress UndefinedInterfaceMethod
     */
    private function addCdnSection(ArrayNodeDefinition $node): void
    {
        $node
            ->children()
                ->arrayNode('cdn')
                    ->children()
                        ->arrayNode('server')
                            ->addDefaultsIfNotSet()
                            ->children()
                                ->scalarNode('path')->defaultValue('/uploads/media')->end()
                            ->end()
                        ->end()

                        ->arrayNode('cloudfront')
                            ->children()
                                ->scalarNode('path')
                                    ->info('e.g. http://xxxxxxxxxxxxxx.cloudfront.net/uploads/media')
                                    ->isRequired()
                                ->end()
                                ->scalarNode('distribution_id')->isRequired()->end()
                                ->scalarNode('key')->isRequired()->end()
                                ->scalarNode('secret')->isRequired()->end()
                                ->scalarNode('region')->isRequired()->end()
                                ->scalarNode('version')->isRequired()->end()
                            ->end()
                        ->end()

                        ->arrayNode('fallback')
                            ->children()
                                ->scalarNode('primary')->isRequired()->end()
                                ->scalarNode('fallback')->isRequired()->end()
                            ->end()
                        ->end()
                    ->end()
                ->end()
            ->end();
    }

    /**
     * @psalm-suppress UndefinedInterfaceMethod
     */
    private function addFilesystemSection(ArrayNodeDefinition $node): void
    {
        // FIXME: add filesystem configuration
    }

    /**
     * @psalm-suppress UndefinedInterfaceMethod
     */
    private function addProvidersSection(ArrayNodeDefinition $node): void
    {
        $node
            ->children()
                ->arrayNode('providers')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->arrayNode('file')
                            ->addDefaultsIfNotSet()
                            ->children()
                                ->scalarNode('service')->defaultValue('sonata.media.provider.file')->end()
                                ->scalarNode('resizer')->defaultNull()->end()
                                ->scalarNode('filesystem')->defaultValue('default.storage')->end()
                                ->scalarNode('cdn')->defaultValue('sonata.media.cdn.server')->end()
                                ->scalarNode('generator')->defaultValue('sonata.media.generator.default')->end()
                                ->scalarNode('thumbnail')->defaultValue('sonata.media.thumbnail.file')->end()
                                ->arrayNode('allowed_extensions')
                                    ->beforeNormalization()
                                    ->ifArray()
                                        ->then(static fn (array $allowedExtensions): array => array_map(strtolower(...), $allowedExtensions))
                                    ->end()
                                    ->prototype('scalar')->end()
                                    ->defaultValue([
                                        'pdf', 'txt', 'rtf',
                                        'doc', 'docx', 'xls', 'xlsx', 'ppt', 'pptx',
                                        'odt', 'odg', 'odp', 'ods', 'odc', 'odf', 'odb',
                                        'csv',
                                        'xml',
                                    ])
                                ->end()
                                ->arrayNode('allowed_mime_types')
                                    ->beforeNormalization()
                                        ->ifArray()
                                        ->then(static fn (array $allowedMimeTypes): array => array_map(strtolower(...), $allowedMimeTypes))
                                    ->end()
                                    ->prototype('scalar')->end()
                                    ->defaultValue([
                                        'application/pdf', 'application/x-pdf', 'application/rtf', 'text/html', 'text/rtf', 'text/plain',
                                        'application/excel', 'application/msword', 'application/vnd.ms-excel', 'application/vnd.ms-powerpoint',
                                        'application/vnd.ms-powerpoint', 'application/vnd.oasis.opendocument.text', 'application/vnd.oasis.opendocument.graphics', 'application/vnd.oasis.opendocument.presentation', 'application/vnd.oasis.opendocument.spreadsheet', 'application/vnd.oasis.opendocument.chart', 'application/vnd.oasis.opendocument.formula', 'application/vnd.oasis.opendocument.database', 'application/vnd.oasis.opendocument.image',
                                        'text/comma-separated-values',
                                        'text/xml', 'application/xml',
                                        'application/zip', // seems to be used for xlsx document ...
                                    ])
                                ->end()
                            ->end()
                        ->end()

                        ->arrayNode('image')
                            ->addDefaultsIfNotSet()
                            ->children()
                                ->scalarNode('service')->defaultValue('sonata.media.provider.image')->end()
                                ->scalarNode('resizer')->defaultValue('sonata.media.resizer.default')->end()
                                ->scalarNode('filesystem')->defaultValue('default.storage')->end()
                                ->scalarNode('cdn')->defaultValue('sonata.media.cdn.server')->end()
                                ->scalarNode('generator')->defaultValue('sonata.media.generator.default')->end()
                                ->scalarNode('thumbnail')->defaultValue('sonata.media.thumbnail.format')->end()
                                ->scalarNode('adapter')->defaultValue('sonata.media.adapter.image.default')->end()
                                ->arrayNode('allowed_extensions')
                                    ->beforeNormalization()
                                        ->ifArray()
                                        ->then(static fn (array $allowedExtensions): array => array_map(strtolower(...), $allowedExtensions))
                                    ->end()
                                    ->prototype('scalar')->end()
                                    ->defaultValue(['jpg', 'png', 'jpeg'])
                                ->end()
                                ->arrayNode('allowed_mime_types')
                                    ->beforeNormalization()
                                        ->ifArray()
                                        ->then(static fn (array $allowedMimeTypes): array => array_map(strtolower(...), $allowedMimeTypes))
                                    ->end()
                                    ->prototype('scalar')->end()
                                    ->defaultValue([
                                        'image/pjpeg',
                                        'image/jpeg',
                                        'image/png',
                                        'image/x-png',
                                    ])
                                ->end()
                            ->end()
                        ->end()

                        ->arrayNode('youtube')
                            ->addDefaultsIfNotSet()
                            ->children()
                                ->scalarNode('service')->defaultValue('sonata.media.provider.youtube')->end()
                                ->scalarNode('resizer')->defaultValue('sonata.media.resizer.default')->end()
                                ->scalarNode('filesystem')->defaultValue('default.storage')->end()
                                ->scalarNode('cdn')->defaultValue('sonata.media.cdn.server')->end()
                                ->scalarNode('generator')->defaultValue('sonata.media.generator.default')->end()
                                ->scalarNode('thumbnail')->defaultValue('sonata.media.thumbnail.format')->end()
                                ->scalarNode('html5')->defaultValue(false)->end()
                            ->end()
                        ->end()

                        ->arrayNode('dailymotion')
                            ->addDefaultsIfNotSet()
                            ->children()
                                ->scalarNode('service')->defaultValue('sonata.media.provider.dailymotion')->end()
                                ->scalarNode('resizer')->defaultValue('sonata.media.resizer.default')->end()
                                ->scalarNode('filesystem')->defaultValue('default.storage')->end()
                                ->scalarNode('cdn')->defaultValue('sonata.media.cdn.server')->end()
                                ->scalarNode('generator')->defaultValue('sonata.media.generator.default')->end()
                                ->scalarNode('thumbnail')->defaultValue('sonata.media.thumbnail.format')->end()
                            ->end()
                        ->end()

                        ->arrayNode('vimeo')
                            ->addDefaultsIfNotSet()
                            ->children()
                                ->scalarNode('service')->defaultValue('sonata.media.provider.vimeo')->end()
                                ->scalarNode('resizer')->defaultValue('sonata.media.resizer.default')->end()
                                ->scalarNode('filesystem')->defaultValue('default.storage')->end()
                                ->scalarNode('cdn')->defaultValue('sonata.media.cdn.server')->end()
                                ->scalarNode('generator')->defaultValue('sonata.media.generator.default')->end()
                                ->scalarNode('thumbnail')->defaultValue('sonata.media.thumbnail.format')->end()
                            ->end()
                        ->end()
                    ->end()
                ->end()
            ->end();
    }

    /**
     * @psalm-suppress UndefinedInterfaceMethod
     */
    private function addModelSection(ArrayNodeDefinition $node): void
    {
        $node
            ->children()
                ->arrayNode('class')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->scalarNode('media')->defaultValue('App\\Entity\\SonataMediaMedia')->end()
                        ->scalarNode('gallery')->defaultValue('App\\Entity\\SonataMediaGallery')->end()
                        ->scalarNode('gallery_item')->defaultValue('App\\Entity\\SonataMediaGalleryItem')->end()
                        ->scalarNode('category')->defaultValue('App\\Entity\\SonataClassificationCategory')->end()
                    ->end()
                ->end()
            ->end();
    }

    /**
     * @psalm-suppress UndefinedInterfaceMethod
     */
    private function addHttpClientSection(ArrayNodeDefinition $node): void
    {
        $node
            ->children()
                ->arrayNode('http')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->scalarNode('client')
                            ->defaultValue('sonata.media.http.base_client')
                            ->info('Alias of the http client.')
                        ->end()
                        ->scalarNode('message_factory')
                            ->defaultValue('sonata.media.http.base_message_factory')
                            ->info('Alias of the message factory.')
                        ->end()
                    ->end()
                ->end()
            ->end();
    }

    /**
     * @psalm-suppress UndefinedInterfaceMethod
     */
    private function addResizerSection(ArrayNodeDefinition $node): void
    {
        $node
            ->children()
                ->arrayNode('resizer')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->arrayNode('simple')
                            ->addDefaultsIfNotSet()
                            ->children()
                                ->integerNode('mode')->defaultValue(ManipulatorInterface::THUMBNAIL_INSET)->end()
                            ->end()
                        ->end()
                        ->arrayNode('square')
                            ->addDefaultsIfNotSet()
                            ->children()
                                ->integerNode('mode')->defaultValue(ManipulatorInterface::THUMBNAIL_INSET)->end()
                            ->end()
                        ->end()
                    ->end()
                ->end()
                ->arrayNode('resizers')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->scalarNode('default')->defaultValue('sonata.media.resizer.simple')->end()
                    ->end()
                ->end()
            ->end();
    }

    /**
     * @psalm-suppress UndefinedInterfaceMethod
     */
    private function addAdapterSection(ArrayNodeDefinition $node): void
    {
        $node
            ->children()
                ->arrayNode('adapters')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->scalarNode('default')->defaultValue('sonata.media.adapter.image.gd')->end()
                    ->end()
                ->end()
            ->end();
    }

    /**
     * @psalm-suppress UndefinedInterfaceMethod
     */
    private function addMessengerSection(ArrayNodeDefinition $node): void
    {
        $node
            ->children()
                ->arrayNode('messenger')
                    ->canBeEnabled()
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->scalarNode('generate_thumbnails_bus')->defaultValue('messenger.default_bus')->end()
                    ->end()
                ->end()
            ->end();
    }
}
